src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -340,4 +348,49 @@ class EntityClient (requestContext: RequestContext, modelSchema: ModelSchema)(im
       }
     }
   }
+
+  def importPFB(pipeline: WithTransformerConcatenation[HttpRequest, Future[HttpResponse]],
+                workspaceNamespace: String, workspaceName: String, pfbRequest: PfbImportRequest): Future[PerRequestMessage] = {
+    def callArrow = {
+      logger.debug("calling arrow")
+      pipeline {
+        Post(avroToRawlsURL, requestContext.request.entity)
+      }
+    }
+
+    def callRawls(arrowResponse: HttpResponse) = {
+      logger.debug("calling rawls")
+      pipeline {
+        Post(FireCloudDirectiveUtils.encodeUri(Rawls.entityPathFromWorkspace(workspaceNamespace, workspaceName) + "/batchUpsert"), arrowResponse.entity)
+      }
+    }
+
+    val pfbUrl = new URL(pfbRequest.url.replace(" ", "%20"))
+    val acceptableProtocols = Seq("https") //for when we inevitably change our mind and need to support others
+
+    if (!acceptableProtocols.contains(pfbUrl.getProtocol)) {
+      Future.successful(RequestCompleteWithErrorReport(StatusCodes.BadRequest, "Invalid URL protocol: must be https only"))
+    } else {
+      val futureResponse = for {
+        arrowResponse <- callArrow
+        rawlsResponse <- arrowResponse.status match {
+          case OK =>
+            callRawls(arrowResponse)
+          case _ =>
+            Future.successful(arrowResponse)
+        }
+      } yield (rawlsResponse.status match {
+        case NoContent =>
+          logger.debug("OK response")
+          RequestComplete(OK)
+        case _ =>
+          // Bubble up all other unmarshallable responses
+          logger.warn("Unanticipated response: " + rawlsResponse.status.defaultMessage)
+          RequestComplete(rawlsResponse)
```

Comment:
consider wrapping this into a `RequestCompleteWithErrorReport` and adding an informative top-level error message. The error message from Rawls may not make sense to an end user trying to import a PFB.

Reply:
I partially agree. Definitely worth saying that the import failed. I still think there's value in including the underlying Rawls message. The easiest way for users to give us error info is to send us a copy of whatever error message we give them. If that included some sort of incident code that we could use to find the relevant logs, that would be really cool. Since we don't have that established practice, the next best thing is to include some detail in the message to hopefully avoid us having to dig around in the logs.

Reply:
I absolutely agree about including the underlying Rawls message; I'm just suggesting wrapping it. What about something like:
```
      } yield rawlsResponse.status match {
        case NoContent =>
          logger.debug("OK response")
          RequestComplete(OK)
        case rawlsStatus =>
          // Bubble up all other unmarshallable responses
          logger.warn("Unanticipated response: " + rawlsResponse.status.defaultMessage)
          // if we can parse Rawls' response as an ErrorReport, wrap it nicely; else just pass it on
          FCErrorReport.tryUnmarshal(rawlsResponse) match {
            case Success(rawlsErrorReport) =>
              RequestCompleteWithErrorReport(rawlsStatus, "Rawls did a baddie; original error is:", Seq(rawlsErrorReport))
            case Failure(_) =>
              RequestComplete(rawlsResponse)
          }
      }
```

Reply:
I like the revamped code better in terms of syntax/Scala-ness. It does not accomplish what I am looking for ... but it may not be necessary. Let me try to explain.

I worry that the error messages from Rawls, which in the current code are bubbled up to the user, may not be helpful. If Rawls returns "503 Service Unavailable" and we pass that to the end user, will the end user know if the problem is in Arrow, Rawls, Orch, or somewhere else? I'd love to wrap the Rawls error in a message like "Successfully retrieved PFB, but encountered an error updating entities. Original error message was: " to help with debugging.

My goal is to help the end user troubleshoot problems, where "end user" might include Workbench devs or QA who encounter test failures. If you are comfortable that the error messages you're bubbling up are enough, then great.